### DISCUSSION TITLE
Prisma doesn't load different .env even with `dotenv-cli`

### DISCUSSION THREAD

**Discussion by joaocasarin** [prisma/prisma#13173] [Q&A]
Well, basically with [this repo](https://github.com/joaocasarin/aws-lambda-typeorm-error/tree/webpack) my prisma migration doesn't load the `.env.test` file I want it to load even when using the command `dotenv -e .env.test -- npx prisma migrate dev --name init`.

This is the output in the console:
![image](https://user-images.githubusercontent.com/48847394/166865564-88bb128f-4ca6-49d1-97d3-8ac9cd46c78e.png)

Pay attention to the part that says `Environment variables loaded from .env`.

I have already tested with different variable values and it is indeed loading from the original `.env` file.

**Comment by janpio** [comment]
`dotenv-cli` should load the env vars from `.env.test` into the environment and then call the Prisma CLI.
`prisma` should load the env vars from `.env` _but not overwrite existing environment variables_.
If that is not the case, this is a bug that can be reported here - optimally with a very simple reproduction we can just run.

**Reply by joaocasarin** [reply]
Hi @janpio , thanks for the quick reply. I don't think I got exactly what you meant, also it doesn't seem that `prisma` is overwriting the existint environment variables. However, it won't read the variables set from the file I chose with the `dotenv-cli`. I'm creating a repository just to reproduce this problem. It's a simple express api with only 1 endpoint, `/create-user`. You think I should open an issue for that?

**Reply by joaocasarin** [reply]
Hey, I have just created this repo to reproduce the "error". Please read the README to see the steps.

Repo: [joaocasarin/dotenv-cli-prisma-error](https://github.com/joaocasarin/dotenv-cli-prisma-error)

PS.: As I'm saying in the README, the prisma is actually loading the correct variables set by `dotenv-cli`, but in the output it is saying wrongly that they were loaded from `.env` instead.

**Reply by janpio** [reply]
Ah, this is only about it showing the wrong message? Sorry, I did not get that in your initial description.

So you are clearly expecting the output to work different than it does. You expect using `dotenv -e .env.test --` should tell Prisma to look in `.env.test` instead of the default file it look into. What is happenign in the code though is that `dotenv -e .env.test --` by itself loads the env variables into the environment, and then Prisma does its thing loading from `.env` but sees that there are already some values and does not overwrite these - hence the values from `.env.test` take precedence.

This is really hard to communicate in the output I fear. (Prisma does know nothing about `.env.test` at all and can not use its name for anything unfortunately. It _only_ sees that env vars are already set.)

Maybe Prisma could print `Loaded "DATABASE_URL" from `.env`, but was already present in environment and did not overwrite". Would that have made it clearer?

**Reply by joaocasarin** [reply]
actually at first it didn't work at all... I was facing `unauthorized` when trying to run the migrations into my docker database, but while creating that repo it started working but still the output was confusing me, so I tried using the `echo` inside the container to see what variables were loaded, and indeed, the variables were loaded from `.env.test`.

Now I get it, I didn't know exactly what `dotenv -e .env.test` would do, thinking that it was just telling prisma to load that file, but instead it was setting the variables to my current environment, and prisma was loading them.
Yea, I get... Since prisma doesn't check the .env file, but environment itself, it is really hard to output the correct .env file....

Mhmm, something like the example output you wrote would prevent future misunderstandings like mine haha

Now, I'm unsure if the problem I was having was about docker, prisma, dotenv or what...

**Comment by patrickkmatias** [comment]
I was facing the same issue like a moments ago in my project and after some time researching I found another package that does practically the same job as `dotenv-cli` does, named `env-cmd`. So the problem is with the package, and not Prisma.

My problem was: I was trying to create a npm script to deploy my migrations in my test db, like this:

`"prisma:test:deploy": "npx dotenv -e ./test/.env.test prisma migrate deploy"`

When it was executed manually it works fine, but I need to run it programmatically in another part of my application, and when it was automatized for some reason it was loading the development .env file (`.env` in my case) with the undesired variables.

So using the `env-cmd` package it worked fine for me even when programmed.

`"prisma:test:deploy": "npx env-cmd -f ./test/.env.test prisma migrate deploy"`

Prisma continues to advertise that is the development `.env` file as reported by you above but what matters is that the right environment variables are being loaded.

**Reply by janpio** [reply]
> Prisma continues to advertise that is the development .env file as reported by you above but what matters is that the right environment variables are being loaded.

I think this might be explainable: Prisma does load the `.env` file - they are just not applied as some values are already present from the earlier command.

**Reply by javodmutalliboev** [reply]
It has really helped me. Thank you. Ma sha Allah.

**Comment by juan-carlos-correa** [comment]
FYI, I was having the same `.env` error.

In my case was because I had in my .env the `DIRECT_URL` variable, but not in my` .env.test `file. 

I just had to add the `DIRECT_URL` to my `.env.test` and it worked.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
